llzk_backend/src/codegen.rs

```diff
@@ -76,39 +170,803 @@ impl<'ast, 'llzk> LlzkCodegen<'ast, 'llzk> {
+            s.gen_llzk_in_function(codegen, &mut func_context)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl GenerateLLZKInModule for TemplateData {
+    fn gen_llzk<'llzk, 'ast: 'llzk>(&'ast self, codegen: &LlzkCodegen<'ast, 'llzk>) -> Result<()> {
+        // Collect declarations first to determine struct fields and function parameters.
+        let mut declarations = DeclarationInfo::default();
+        for s in self.get_body_as_vec() {
+            declarations.visit(codegen, s)?;
+        }
+
+        // Generate the struct definition, prepopulated with fields.
+        let struct_loc = codegen.location(self.get_file_id(), self.get_param_location());
+        let struct_params: Vec<_> = self.get_name_of_params().iter().map(String::as_str).collect();
+        let struct_def =
+            r#struct::def(struct_loc, self.get_name(), &struct_params, declarations.struct_fields)?;
+        let new_struct = codegen.add_struct(struct_def)?;
+
+        // Generate the compute and constrain functions.
+        let inputs: Vec<_> = declarations.inputs.iter().map(|v| v.type_and_loc).collect();
+        let arg_attrs: Vec<_> = declarations.inputs.iter().map(|v| v.attrs.as_slice()).collect();
+        let new_struct_type = new_struct.r#type();
+        let struct_body = new_struct.body();
+        let compute_func = FuncDefOpRef::try_from(struct_body.append_operation(
+            compute_fn(struct_loc, new_struct_type, &inputs, Some(&arg_attrs))?.into(),
+        ))?
+        .into();
+        let constrain_func = FuncDefOpRef::try_from(struct_body.append_operation(
+            constrain_fn(struct_loc, new_struct_type, &inputs, Some(&arg_attrs))?.into(),
+        ))?
+        .into();
+
+        // Map parameter Values of each LLZK function to the corresponding circom variable names and
+        // then create the FunctionContext for each function. Before creating the FunctionContext
+        // for constrain, add a dummy name at index 0 since the first parameter is the struct ref.
+        let mut arg_names: Vec<_> = declarations.inputs.iter().map(|i| i.name.clone()).collect();
+        let mut compute_ctx =
+            FunctionContext::new(compute_func, map_name_to_arg_value(compute_func, &arg_names)?)?;
+        arg_names.insert(0, "**self**".to_string());
+        let mut constrain_ctx = FunctionContext::new(
+            constrain_func,
+            map_name_to_arg_value(constrain_func, &arg_names)?,
+        )?;
+        // Insert the Operations created from variable Declaration statements and map the circom
+        // variable name to LLVM op result Value (do this in each function).
+        declarations.var_decls.into_iter().for_each(|(name, op)| {
+            // Insert (a clone of) the declaration into the compute function.
+            compute_ctx.append_op_named_result(op.clone(), name.clone());
+            // Insert the declaration into the constrain function.
+            constrain_ctx.append_op_named_result(op, name);
+        });
```

Comment:
A bit cleaner (if it is possible since there's a partial move I don't know if this compiles just by looking)
```suggestion
        for (name, op) in declarations.var_decls {
            // Insert (a clone of) the declaration into the compute function.
            compute_ctx.append_op_named_result(op.clone(), name.clone());
            // Insert the declaration into the constrain function.
            constrain_ctx.append_op_named_result(op, name);
        }
```